File: hermes/tools/zoho_crm_tool.py
"""Hermes-native Zoho CRM model tool."""

from __future__ import annotations

from typing import Any

from tools.registry import registry, tool_error, tool_result
from tools.zoho_approval import require_zoho_write_approval
from tools.zoho_runtime import zoho_tool_available
from tools.zoho_client import (
    DEFAULT_INLINE_THRESHOLD,
    ZohoClient,
    build_deal_forecast,
    build_lead_report,
    build_pipeline_summary,
    format_crm_result,
    normalize_crm_module,
    parse_date_filter,
    records_to_csv,
)


CRM_READ_OPS = {
    "list",
    "get",
    "search",
    "search_text",
    "build_pipeline_summary",
    "build_lead_report",
    "build_deal_forecast",
}
CRM_WRITE_OPS = {"create", "update", "delete"}
CRM_MODULES = ["Leads", "Contacts", "Accounts", "Deals", "Tasks"]


ZOHO_CRM_SCHEMA = {
    "name": "zoho_crm",
    "description": (
        "Access Zoho CRM records and reports. Supports Divo-compatible "
        "operations for Leads, Contacts, Accounts, Deals, and Tasks: list, get, "
        "criteria search, free-text search, create, update, delete, pipeline "
        "summary, lead report, and deal forecast."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "op": {
                "type": "string",
                "enum": sorted(CRM_READ_OPS | CRM_WRITE_OPS),
                "description": "Divo-compatible operation name.",
            },
            "operation": {
                "type": "string",
                "description": "Alias for op. Prefer op for Divo compatibility.",
            },
            "module": {
                "type": "string",
                "enum": CRM_MODULES + [
                    "lead",
                    "contact",
                    "account",
                    "deal",
                    "task",
                    "opportunity",
                    "company",
                    "customer",
                ],
                "description": "CRM module for CRUD/search operations.",
            },
            "recordId": {"type": "string"},
            "criteria": {
                "type": "string",
                "description": "Zoho CRM criteria syntax, e.g. (Last_Name:contains:Smith).",
            },
            "query": {
                "type": "string",
                "description": "Free-text search query for search_text.",
            },
            "fields": {
                "type": "object",
                "additionalProperties": True,
                "description": "Zoho CRM record fields for create/update.",
            },
            "limit": {
                "type": "integer",
                "minimum": 1,
                "maximum": 200,
                "description": "Maximum inline records to return. Default 25.",
            },
            "sortBy": {"type": "string"},
            "sortOrder": {"type": "string", "enum": ["asc", "desc"]},
            "exportAll": {
                "type": "boolean",
                "description": "Exhaust all pages and include summary metadata.",
            },
            "exportCsv": {
                "type": "boolean",
                "description": "Include an inline CSV string for returned/exported rows.",
            },
            "csvColumns": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Optional CSV column order.",
            },
            "closingFrom": {"type": "string"},
            "closingTo": {"type": "string"},
            "currency": {"type": "string"},
        },
        "required": ["op"],
    },
}


async def _handle_zoho_crm(args: dict[str, Any], **kwargs: Any) -> str:
    op = _operation(args)
    if not op:
        return tool_error("op is required", success=False)
    if op not in CRM_READ_OPS and op not in CRM_WRITE_OPS:
        return tool_error(f"Unknown Zoho CRM operation: {op}", success=False)

    try:
        from tools.zoho_runtime import resolve_tool_client

        client = resolve_tool_client(kwargs)
    except Exception as exc:
        return tool_error(_map_zoho_error(exc), success=False, operation=op)

    try:
        if op in CRM_WRITE_OPS:
            approved = require_zoho_write_approval(
                pattern_key=f"zoho_crm:{op}",
                action=f"zoho_crm {op}",
                description=_write_description(op, args),
                approval_callback=kwargs.get("approval_callback"),
            )
            if not approved.get("approved"):
                return tool_error(
                    approved.get("message") or "Zoho CRM write approval required.",
                    success=False,
                    operation=op,
                    status=approved.get("status", "blocked"),
                    approval_pending=approved.get("approval_pending", False),
                    pattern_key=approved.get("pattern_key"),
                )
            result = await _execute_write(client, op, args)
            result["approval"] = approved.get("approval")
            return tool_result(result)

        result = await _execute_read(client, op, args)
        return tool_result(result)
    except Exception as exc:
        return tool_error(_map_zoho_error(exc), success=False, operation=op)


async def _execute_read(
    client: ZohoClient,
    op: str,
    args: dict[str, Any],
) -> dict[str, Any]:
    if op == "list":
        module = _module(args, op)
        if bool(args.get("exportAll")):
            result = await client.crm_list_all_records(
                module,
                sort_by=_str_arg(args, "sortBy"),
                sort_order=_sort_order(args),
            )
            items = result["items"]
            inline = items[: _limit(args)]
            csv_text = _maybe_csv(args, items)
            message = f"Found {len(items)} {normalize_crm_module(module)} record(s)."
            if len(items) > len(inline):
                message += f" Showing {len(inline)} inline."
            if csv_text:
                message += " Inline CSV included."
            if result.get("truncated"):
                message += " Pagination limit reached; additional records may exist."
            return {
                "success": True,
                "data": format_crm_result(inline),
                "message": message,
                "truncated": bool(result.get("truncated")),
                "hasMore": len(items) > len(inline) or bool(result.get("truncated")),
                "report": {
                    "items": format_crm_result(inline),
                    "totalCount": len(items),
                    "truncated": bool(result.get("truncated")),
                    **({"csv": csv_text} if csv_text else {}),
                },
                **({"csv": csv_text} if csv_text else {}),
            }

        result = await client.crm_list_records(
            module,
            per_page=_limit(args),
            sort_by=_str_arg(args, "sortBy"),
            sort_order=_sort_order(args),
        )
        items = result["items"]
        return {
            "success": True,
            "data": format_crm_result(items),
            "message": f"Found {len(items)} {normalize_crm_module(module)} record(s).",
            "hasMore": bool(result.get("hasMore")),
        }

    if op == "get":
        module = _module(args, op)
        record_id = _required(args, "recordId", op)
        record = await client.crm_get_record(module, record_id)
        return {
            "success": True,
            "data": format_crm_result(record),
            "message": "Record not found" if record is None else None,
        }

    if op == "search":
        module = _module(args, op)
        criteria = _required(args, "criteria", op)
        result = await client.crm_search_records(
            module,
            criteria,
            per_page=_limit(args),
        )
        items = result["items"]
        mod = normalize_crm_module(module)
        return {
            "success": True,
            "data": format_crm_result(items),
            "message": (
                f"Found {len(items)} {mod} record(s)."
                if items
                else f"No {mod} records matched the search criteria."
            ),
            "hasMore": bool(result.get("hasMore")),
        }

    if op == "search_text":
        module = _module(args, op)
        query = _required(args, "query", op)
        result = await client.crm_search_by_text(
            module,
            query,
            per_page=_limit(args),
        )
        items = result["items"]
        mod = normalize_crm_module(module)
        return {
            "success": True,
            "data": format_crm_result(items),
            "message": (
                f"Found {len(items)} {mod} record(s) matching \"{query}\"."
                if items
                else f"No {mod} records found matching \"{query}\"."
            ),
            "hasMore": bool(result.get("hasMore")),
        }

    if op == "build_pipeline_summary":
        report = await build_pipeline_summary(
            client,
            currency=_str_arg(args, "currency") or "INR",
            inline_threshold=min(_limit(args), 200),
        )
        return {
            "success": True,
            "message": report["summary"],
            "report": format_crm_result(report),
        }

    if op == "build_lead_report":
        report = await build_lead_report(client, inline_threshold=min(_limit(args), 200))
        return {
            "success": True,
            "message": report["summary"],
            "report": format_crm_result(report),
        }

    if op == "build_deal_forecast":
        closing_from = _date_bound(args, "closingFrom", "from")
        closing_to = _date_bound(args, "closingTo", "to")
        report = await build_deal_forecast(
            client,
            closing_from=closing_from,
            closing_to=closing_to,
            currency=_str_arg(args, "currency") or "INR",
            inline_threshold=min(_limit(args), 200),
        )
        return {
            "success": True,
            "message": report["summary"],
            "report": format_crm_result(report),
        }

    raise ValueError(f"Unhandled Zoho CRM read operation: {op}")


async def _execute_write(
    client: ZohoClient,
    op: str,
    args: dict[str, Any],
) -> dict[str, Any]:
    module = _module(args, op)
    mod = normalize_crm_module(module)

    if op == "create":
        fields = _fields(args, op)
        result = await client.crm_create_record(module, fields)
        return {
            "success": True,
            "recordId": result["id"],
            "message": f"{mod} record created",
            "data": format_crm_result(result.get("data")),
        }

    if op == "update":
        record_id = _required(args, "recordId", op)
        fields = _fields(args, op)
        result = await client.crm_update_record(module, record_id, fields)
        return {
            "success": True,
            "recordId": record_id,
            "message": f"{mod} record updated",
            "data": format_crm_result(result),
        }

    if op == "delete":
        record_id = _required(args, "recordId", op)
        result = await client.crm_delete_record(module, record_id)
        return {
            "success": True,
            "recordId": record_id,
            "message": f"{mod} record deleted",
            "data": format_crm_result(result),
        }

    raise ValueError(f"Unhandled Zoho CRM write operation: {op}")


def _operation(args: dict[str, Any]) -> str:
    return str(args.get("op") or args.get("operation") or "").strip()


def _limit(args: dict[str, Any]) -> int:
    try:
        return max(1, min(200, int(args.get("limit") or DEFAULT_INLINE_THRESHOLD)))
    except (TypeError, ValueError):
        return DEFAULT_INLINE_THRESHOLD


def _module(args: dict[str, Any], op: str) -> str:
    module = _str_arg(args, "module")
    if not module:
        raise ValueError(f"module is required for {op}")
    return module


def _sort_order(args: dict[str, Any]) -> str | None:
    value = _str_arg(args, "sortOrder")
    return value if value in {"asc", "desc"} else None


def _date_bound(args: dict[str, Any], key: str, side: str) -> str | None:
    value = _str_arg(args, key)
    if not value:
        return None
    return parse_date_filter(value)[side]


def _maybe_csv(args: dict[str, Any], items: list[dict[str, Any]]) -> str | None:
    if not (bool(args.get("exportCsv")) or bool(args.get("exportAll"))):
        return None
    columns = args.get("csvColumns")
    return records_to_csv(items, columns if isinstance(columns, list) else None)


def _str_arg(args: dict[str, Any], key: str) -> str | None:
    value = args.get(key)
    return value.strip() if isinstance(value, str) and value.strip() else None


def _required(args: dict[str, Any], key: str, op: str) -> str:
    value = _str_arg(args, key)
    if not value:
        raise ValueError(f"{key} is required for {op}")
    return value


def _fields(args: dict[str, Any], op: str) -> dict[str, Any]:
    fields = args.get("fields")
    if not isinstance(fields, dict) or not fields:
        raise ValueError(f"fields is required for {op}")
    return fields


def _write_description(op: str, args: dict[str, Any]) -> str:
    identifiers = []
    module = _str_arg(args, "module")
    record_id = _str_arg(args, "recordId")
    if module:
        identifiers.append(f"module={normalize_crm_module(module)}")
    if record_id:
        identifiers.append(f"recordId={record_id}")
    detail = f" ({', '.join(identifiers)})" if identifiers else ""
    return f"Zoho CRM write operation '{op}'{detail}"


def _map_zoho_error(exc: Exception) -> str:
    text = str(exc)
    lower = text.lower()
    if "1002" in text or "invalid oauth" in lower or "unauthorized" in lower:
        return "Zoho authentication failed. Reconnect Zoho or refresh credentials."
    if "invalid module" in lower:
        return "Zoho CRM module is invalid. Use Leads, Contacts, Accounts, Deals, or Tasks."
    return text


registry.register(
    name="zoho_crm",
    toolset="zoho",
    schema=ZOHO_CRM_SCHEMA,
    handler=_handle_zoho_crm,
    check_fn=zoho_tool_available,
    requires_env=["ZOHO_CLIENT_ID", "ZOHO_CLIENT_SECRET", "ZOHO_REFRESH_TOKEN"],
    is_async=True,
    emoji="🏢",
    max_result_size_chars=100_000,
)

File: advance-backend/src/application/tools/families/zoho-crm.tool.ts
/**
 * zohoCrm tool — Zoho CRM read/write + pipeline & lead reports.
 *
 * Operations:
 *   CRUD (generic — works with any module: Leads, Contacts, Accounts, Deals, Tasks):
 *     list           — paginated record list
 *     get            — single record by ID
 *     search         — criteria-based search (Zoho CRM criteria syntax)
 *     search_text    — free-text search across name/email fields
 *     create         — create a record
 *     update         — update a record
 *     delete         — delete a record
 *
 *   Reports (exhaustive pagination + token-safe output):
 *     build_pipeline_summary — deals by stage with amounts
 *     build_lead_report      — lead funnel by source/status
 *     build_deal_forecast    — deals closing within a date range
 *
 * Result safety:
 *   - CRUD ops return at most `limit` records (default 25, max 200)
 *   - Reports return a summary plus a bounded inline sample
 */

import { z } from 'zod';
import type { Tool, ToolExecutionContext } from '../tool.contract';
import type { Result }                     from '../../../shared/result';
import { ok, err }                         from '../../../shared/result';
import { PermissionError, ToolError }      from '../../../shared/errors';
import type { ToolActionGroup }            from '../../../domain/permissions/tool-action-group';
import { asToolId }                        from '../../../shared/ids';
import type { ZohoCrmOps }                 from '../../zoho/zoho-crm-ops';
import { mapZohoError }                    from '../../zoho/zoho-error.utils';
import type { ZohoCrmPaginatedClient } from '../../../infrastructure/zoho/zoho-crm-paginated.client';
import { parseDateFilter } from '../../zoho/zoho-filter.utils';
import { filterZohoRecordsByEmail, normalizedEmail, recordMatchesZohoEmail } from '../../../shared/zoho-personalization';

// ─── Args schema ──────────────────────────────────────────────────────────────

const connectionField = { connectionId: z.string().uuid().optional() } as const;
const moduleField = { module: z.string().min(1) } as const;
const pageField = {
  limit: z.number().int().min(1).max(200).optional(),
  page: z.number().int().min(1).max(10).optional(),
} as const;

const ListSchema = z.object({
  ...connectionField,
  ...moduleField,
  ...pageField,
  op: z.literal('list'),
  pageToken: z.string().min(1).max(2048).optional(),
  sortBy: z.string().optional(),
  sortOrder: z.enum(['asc', 'desc']).optional(),
}).strict();

const Schema = z.discriminatedUnion('op', [
  ListSchema,
  z.object({ ...connectionField, ...moduleField, op: z.literal('get'), recordId: z.string().min(1) }).strict(),
  z.object({ ...connectionField, ...moduleField, ...pageField, op: z.literal('search'), criteria: z.string().min(1) }).strict(),
  z.object({ ...connectionField, ...moduleField, ...pageField, op: z.literal('search_text'), query: z.string().min(1) }).strict(),
  z.object({ ...connectionField, ...moduleField, op: z.literal('create'), fields: z.record(z.unknown()) }).strict(),
  z.object({ ...connectionField, ...moduleField, op: z.literal('update'), recordId: z.string().min(1), fields: z.record(z.unknown()) }).strict(),
  z.object({ ...connectionField, ...moduleField, op: z.literal('delete'), recordId: z.string().min(1) }).strict(),
  z.object({ ...connectionField, op: z.literal('build_pipeline_summary') }).strict(),
  z.object({ ...connectionField, op: z.literal('build_lead_report') }).strict(),
  z.object({
    ...connectionField,
    op: z.literal('build_deal_forecast'),
    closingFrom: z.string().optional(),
    closingTo: z.string().optional(),
  }).strict(),
]).superRefine((value, ctx) => {
  if (value.op === 'list' && value.page !== undefined && value.pageToken !== undefined) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['pageToken'],
      message: 'pageToken cannot be combined with page',
    });
  }
});

type Args = z.infer<typeof Schema>;

const ResultSchema = z.object({
  success:         z.boolean(),
  data:            z.unknown().optional(),
  recordId:        z.string().optional(),
  message:         z.string().optional(),
  report:          z.unknown().optional(),
  truncated:       z.boolean().optional(),
  hasMore:         z.boolean().optional(),
  page:            z.number().int().positive().optional(),
  nextPage:        z.number().int().positive().optional(),
  nextPageToken:   z.string().optional(),
});

type Res = z.infer<typeof ResultSchema>;

// ─── Constants ────────────────────────────────────────────────────────────────

const readOps = new Set<Args['op']>([
  'list', 'get', 'search', 'search_text',
  'build_pipeline_summary', 'build_lead_report', 'build_deal_forecast',
]);

export const zohoCrmActionFor = (op: string): ToolActionGroup =>
  readOps.has(op as Args['op']) ? 'read'
    : op === 'create' ? 'create'
      : op === 'update' ? 'update'
        : 'delete';

const isRecord = (value: unknown): value is Record<string, unknown> =>
  value !== null && typeof value === 'object' && !Array.isArray(value);

function formatCrmResult(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(formatCrmResult);
  if (!isRecord(value)) return value;

  const formatted: Record<string, unknown> = {};
  for (const [key, fieldValue] of Object.entries(value)) {
    if (isRecord(fieldValue) && 'id' in fieldValue && 'name' in fieldValue) {
      formatted[key] = fieldValue['name'];
      formatted[`${key}_id`] = fieldValue['id'];
    } else {
      formatted[key] = formatCrmResult(fieldValue);
    }
  }
  return formatted;
}

// ─── Tool factory ─────────────────────────────────────────────────────────────

export const createZohoCrmTool = (deps: {
  crmClient:  ZohoCrmPaginatedClient;
  crmOps:     ZohoCrmOps;
}): Tool<Args, Res> => ({
  id:           asToolId('zohoCrm'),
  family:       'zoho',
  actionGroups: new Set(['read', 'create', 'update', 'delete']),
  argsSchema:   Schema,
  resultSchema: ResultSchema,

  description: [
    'Access Zoho CRM: list, get, search, create, update, delete records across Leads, Contacts, Accounts, Deals, Tasks.',
    'Search uses Zoho criteria syntax: (Field:operator:value) with and/or.',
    'Reports: build_pipeline_summary (deals by stage), build_lead_report (funnel by source), build_deal_forecast (closing deals).',
    'For a complete list workflow, page from a governed local script using page/nextPage, then pageToken when returned.',
  ].join(' '),

  parameterDocs: [
    'connectionId: exact accessible Zoho UUID. In backend-hosted channels, omit it when only one Zoho account is accessible; the backend resolves that account. If multiple are available, retry with the exact ID returned by the error.',
    'op: list|get|search|search_text|create|update|delete|build_pipeline_summary|build_lead_report|build_deal_forecast',
    'module: Leads|Contacts|Accounts|Deals|Tasks (required for CRUD ops)',
    'recordId: record ID (required for get/update/delete)',
    'criteria: Zoho search criteria (for search op). Format: (Field:operator:value)and/or(Field:operator:value)',
    '  Operators: equals, starts_with, contains, not_equal, greater_than, less_than, between',
    '  Example: "(Deal_Name:contains:Acme)and(Stage:equals:Qualification)"',
    'query: free-text search (for search_text op) — searches name/email fields',
    'fields: record fields for create/update',
    'limit: max records to return (1-200, default 25 direct or 200 in a local-file workflow)',
    'page: list/search page 1-10. When hasMore=true, call nextPage; list switches to nextPageToken when Zoho returns one.',
    'pageToken: opaque continuation returned by a prior list call; do not combine with page.',
    'sortBy: field name to sort by (e.g., Created_Time, Amount)',
    'sortOrder: asc|desc',
    '',
    'REPORT PARAMS:',
    'build_deal_forecast: closingFrom, closingTo (ISO dates or natural: "this month", "this quarter")',
    '',
    'CRM MODULE FIELDS:',
    'Leads: First_Name, Last_Name, Email, Company, Phone, Lead_Source, Lead_Status, Annual_Revenue, Owner',
    'Contacts: First_Name, Last_Name, Email, Phone, Account_Name (lookup), Owner',
    'Accounts: Account_Name, Website, Phone, Industry, Annual_Revenue, Account_Type, Owner',
    'Deals: Deal_Name, Amount, Stage, Closing_Date, Account_Name (lookup), Contact_Name (lookup), Probability, Owner',
    'Tasks: Subject, Due_Date, Status, Priority, Who_Id (contact lookup), What_Id (deal/account lookup), Owner',
  ].join('\n'),

  permissionCheck(args, perm) {
    const action = zohoCrmActionFor(args.op);
    const allowed = perm.allowedActionsByTool.get(asToolId('zohoCrm'))?.has(action) ?? false;
    return allowed ? ok(action) : err(new PermissionError({ toolId: 'zohoCrm', action, reason: 'not_allowed' }));
  },

  async execute(args: Args, ctx: ToolExecutionContext): Promise<Result<Res, ToolError>> {
    const { companyId, userId } = ctx.runContext;
    const connectionContext = {
      userId,
      ...(args.connectionId ? { connectionId: args.connectionId } : {}),
    };
    const personalizedScope = ctx.perm.department?.zohoReadScope === 'personalized';
    const requesterEmail = normalizedEmail(ctx.runContext.requesterEmail);
    if (personalizedScope && !requesterEmail) {
      return err(new ToolError({
        toolId: 'zohoCrm',
        reason: 'permission_denied',
        message: 'Personalized Zoho access requires the signed-in member email.',
      }));
    }
    if (personalizedScope && !readOps.has(args.op)) {
      return err(new ToolError({
        toolId: 'zohoCrm',
        reason: 'permission_denied',
        message: 'Zoho write actions are unavailable while this role is restricted to personalized data.',
      }));
    }
    if (personalizedScope) {
      ctx.logger.info('zoho_crm.scope.personalized', { requesterEmail, op: args.op });
    }

    // ── Report operations ─────────────────────────────────────────────────────
    if (args.op === 'build_pipeline_summary') {
      if (personalizedScope) return err(new ToolError({ toolId: 'zohoCrm', reason: 'permission_denied', message: 'Pipeline summaries are unavailable for personalized Zoho access.' }));
      ctx.onProgress?.('Building CRM pipeline summary…');
      try {
        const report = await deps.crmOps.buildPipelineSummary({ companyId, ...connectionContext });
        return ok({
          success: true,
          message: report.summary,
          report:  formatCrmResult(report),
        });
      } catch (e) {
        return err(new ToolError({
          toolId: 'zohoCrm', reason: 'upstream_failure', cause: e,
          message: `Pipeline summary failed: ${mapZohoError(e)}`,
        }));
      }
    }

    if (args.op === 'build_lead_report') {
      if (personalizedScope) return err(new ToolError({ toolId: 'zohoCrm', reason: 'permission_denied', message: 'Lead reports are unavailable for personalized Zoho access.' }));
      ctx.onProgress?.('Building CRM lead report…');
      try {
        const report = await deps.crmOps.buildLeadReport({ companyId, ...connectionContext });
        return ok({
          success: true,
          message: report.summary,
          report:  formatCrmResult(report),
        });
      } catch (e) {
        return err(new ToolError({
          toolId: 'zohoCrm', reason: 'upstream_failure', cause: e,
          message: `Lead report failed: ${mapZohoError(e)}`,
        }));
      }
    }

    if (args.op === 'build_deal_forecast') {
      if (personalizedScope) return err(new ToolError({ toolId: 'zohoCrm', reason: 'permission_denied', message: 'Deal forecasts are unavailable for personalized Zoho access.' }));
      ctx.onProgress?.('Building deal forecast…');
      try {
        const closingFrom = args.closingFrom ? parseDateFilter(args.closingFrom).from : undefined;
        const closingTo   = args.closingTo   ? parseDateFilter(args.closingTo).to     : undefined;
        const report = await deps.crmOps.buildDealForecast({
          companyId,
          ...connectionContext,
          ...(closingFrom ? { closingFrom } : {}),
          ...(closingTo ? { closingTo } : {}),
        });
        return ok({
          success: true,
          message: report.summary,
          report:  formatCrmResult(report),
        });
      } catch (e) {
        return err(new ToolError({
          toolId: 'zohoCrm', reason: 'upstream_failure', cause: e,
          message: `Deal forecast failed: ${mapZohoError(e)}`,
        }));
      }
    }

    // ── CRUD operations via paginated client ──────────────────────────────────
    const mod = args.module;

    try {
      switch (args.op) {
        case 'list': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for list' }));
          ctx.onProgress?.(`Listing ${mod}…`);

          const result = await deps.crmClient.listRecords({
            companyId, ...connectionContext, module: mod,
            perPage: args.limit ?? (ctx.resultAudience === 'local_file' ? 200 : 25),
            ...(args.pageToken ? { pageToken: args.pageToken } : { page: args.page ?? 1 }),
            ...(args.sortBy ? { sortBy: args.sortBy } : {}),
            ...(args.sortOrder ? { sortOrder: args.sortOrder } : {}),
          });

          const items = personalizedScope ? filterZohoRecordsByEmail(result.items, requesterEmail!) : result.items;
          const continuationMissing = result.hasMore
            && !result.nextPageToken
            && (result.page === undefined || result.page >= 10);
          return ok({
            success: true,
            data: formatCrmResult(items),
            message: continuationMissing
              ? `Found ${items.length} ${mod} record(s). Zoho reported more records but returned no continuation token.`
              : `Found ${items.length} ${mod} record(s).`,
            hasMore: result.hasMore,
            ...(continuationMissing ? { truncated: true } : {}),
            ...(result.page !== undefined ? { page: result.page } : {}),
            ...(result.hasMore && result.page !== undefined && result.page < 10 && !result.nextPageToken
              ? { nextPage: result.page + 1 }
              : {}),
            ...(result.nextPageToken ? { nextPageToken: result.nextPageToken } : {}),
          });
        }

        case 'get': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for get' }));
          if (!args.recordId) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'recordId is required for get' }));
          ctx.onProgress?.(`Fetching ${mod} record…`);
          const record = await deps.crmClient.getRecord({ companyId, ...connectionContext, module: mod, recordId: args.recordId });
          if (!record || (personalizedScope && !recordMatchesZohoEmail(record, requesterEmail!))) return ok({ success: true, data: null, message: 'Record not found' });
          return ok({ success: true, data: formatCrmResult(record) });
        }

        case 'search': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for search' }));
          if (!args.criteria) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'criteria is required for search' }));
          ctx.onProgress?.(`Searching ${mod}…`);
          const result = await deps.crmClient.searchRecords({
            companyId, ...connectionContext, module: mod,
            criteria: args.criteria,
            perPage: args.limit ?? (ctx.resultAudience === 'local_file' ? 200 : 25),
            page: args.page ?? 1,
          });
          const items = personalizedScope ? filterZohoRecordsByEmail(result.items, requesterEmail!) : result.items;
          const sourceTruncated = result.hasMore && (result.page ?? args.page ?? 1) >= 10;
          return ok({
            success: true,
            data: formatCrmResult(items),
            message: items.length > 0
              ? `Found ${items.length} ${mod} record(s).${sourceTruncated ? ' Zoho search reached its 2,000-record limit.' : ''}`
              : `No ${mod} records matched the search criteria.`,
            hasMore: result.hasMore,
            ...(result.page !== undefined ? { page: result.page } : {}),
            ...(result.hasMore && !sourceTruncated ? { nextPage: (result.page ?? 1) + 1 } : {}),
            ...(sourceTruncated ? { truncated: true } : {}),
          });
        }

        case 'search_text': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for search_text' }));
          if (!args.query) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'query is required for search_text' }));
          ctx.onProgress?.(`Searching ${mod} for "${args.query}"…`);
          const result = await deps.crmClient.searchByText({
            companyId, ...connectionContext, module: mod,
            query: args.query,
            perPage: args.limit ?? (ctx.resultAudience === 'local_file' ? 200 : 25),
            page: args.page ?? 1,
          });
          const items = personalizedScope ? filterZohoRecordsByEmail(result.items, requesterEmail!) : result.items;
          const sourceTruncated = result.hasMore && (result.page ?? args.page ?? 1) >= 10;
          return ok({
            success: true,
            data: formatCrmResult(items),
            message: items.length > 0
              ? `Found ${items.length} ${mod} record(s) matching "${args.query}".${sourceTruncated ? ' Zoho search reached its 2,000-record limit.' : ''}`
              : `No ${mod} records found matching "${args.query}".`,
            hasMore: result.hasMore,
            ...(result.page !== undefined ? { page: result.page } : {}),
            ...(result.hasMore && !sourceTruncated ? { nextPage: (result.page ?? 1) + 1 } : {}),
            ...(sourceTruncated ? { truncated: true } : {}),
          });
        }

        case 'create': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for create' }));
          if (!args.fields) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'fields is required for create' }));
          ctx.onProgress?.(`Creating ${mod} record…`);
          const result = await deps.crmClient.createRecord({ companyId, ...connectionContext, module: mod, fields: args.fields });
          return ok({ success: true, recordId: result.id, message: `${mod} record created` });
        }

        case 'update': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for update' }));
          if (!args.recordId) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'recordId is required for update' }));
          if (!args.fields) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'fields is required for update' }));
          ctx.onProgress?.(`Updating ${mod} record…`);
          await deps.crmClient.updateRecord({ companyId, ...connectionContext, module: mod, recordId: args.recordId, fields: args.fields });
          return ok({ success: true, recordId: args.recordId, message: `${mod} record updated` });
        }

        case 'delete': {
          if (!mod) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'module is required for delete' }));
          if (!args.recordId) return err(new ToolError({ toolId: 'zohoCrm', reason: 'bad_args', message: 'recordId is required for delete' }));
          ctx.onProgress?.(`Deleting ${mod} record…`);
          await deps.crmClient.deleteRecord({ companyId, ...connectionContext, module: mod, recordId: args.recordId });
          return ok({ success: true, recordId: args.recordId, message: `${mod} record deleted` });
        }
      }
    } catch (e) {
      return err(new ToolError({
        toolId:  'zohoCrm',
        reason:  'upstream_failure',
        cause:   e,
        message: mapZohoError(e),
      }));
    }
  },
});
